### ISSUE TITLE
OneHotEncoder incorrectly encodes -1 when use_cat_names is True

### ISSUE THREAD

**Issue by jvs** [scikit-learn-contrib/categorical-encoding#132]
When `use_cat_names` is True, OneHotEncoder incorrectly encodes values that equal -1.

Here's a minimal example:
```python
from category_encoders.one_hot import OneHotEncoder

encoder1 = OneHotEncoder(cols=[0], use_cat_names=False)
encoder2 = OneHotEncoder(cols=[0], use_cat_names=True)

a = encoder1.fit_transform([[-1]])
b = encoder2.fit_transform([[-1]])

print('a =', a.values)
print('b =', b.values)
```

This outputs:
```
a = [[1 0]]
b = [[0 0]]
```

I would expect `b` to have the same values as `a`.

**Comment by janmotl** [comment]
Thank you. The treatment of missing values, which should be represented by -1, is truly broken.

**Comment by janmotl** [comment]
There is a clash of generated names in `get_dummies()`. 

Specifically, when we have all the following:
1. `handle_unknown == 'impute'`
2. `use_cat_names == True` 
3. the input contains `-1`

Then:
1. one column name with `-1` suffix gets created because `handle_unknown == 'impute'` 
2. and another column name with `-1` suffix gets created because `use_cat_names == True` and the input contains `-1`

And DataFrame, quite understandably, does not support two columns with the same name.

Possible solutions:
1. find a sentinel value that can never appear in the input data and use it to mark imputed columns
2. write a function that generates unique column names (e.g.: We may keep adding '#' to the end of the name, until we generate a name that has not been observed before. This can be implemented with a reasonable overhead by using a set to keep track of the seen names.)

I am not particularly happy with either approach. Some opinions?

**Comment by janmotl** [comment]
I was wrong. It is possible to create a DataFrame with duplicate column names:
```python
    def test_one_hot_cat_names(self):
        # currently both these combinations will result into columns named 'Apple_A_1'
        X = pd.DataFrame({'Apple_A': ['1'], 'Apple': ['A_1']})
        enc = encoders.OneHotEncoder(use_cat_names=True)
        result = enc.fit_transform(X)

        self.assertEqual(len(result.columns.unique()), len(result.columns), 'Columns must have unique names')
```
It's just inconvenient to use the result (e.g.: indexing by name does not work). 

So, I see two options:
1. return a DataFrame with duplicate column names
2. write a function, which generates unique names

I prefer the second approach. Go ahead.

**Comment by janmotl** [comment]
Solved with #145
